src/NumericalAlgorithms/Interpolation/MultiLinearSpanInterpolation.hpp

```diff
@@ -0,0 +1,397 @@
+  };
+
+  template <size_t... VariablesToInterpolate>
+  std::array<double, sizeof...(VariablesToInterpolate)> interpolate(
+      weight_t<Dimension> const& weights) const {
+    static_assert(sizeof...(VariablesToInterpolate) <= NumberOfVariables,
+                  "You are trying to interpolate more variables than this "
+                  "container holds.");
+
+    return std::array<double, sizeof...(VariablesToInterpolate)>{
+        interpolate(weights, VariablesToInterpolate)...};
+  }
+
+  template <size_t NumberOfVariablesToInterpolate, typename... T,
+            Requires<(std::is_floating_point_v<typename std::remove_cv_t<T>> and
+                      ...)> = nullptr>
+  std::array<double, NumberOfVariablesToInterpolate> interpolate(
+      Index<NumberOfVariablesToInterpolate>& VariablesToInterpolate,
+      T const&... targetPoints) const {
+    static_assert(
+        sizeof...(T) == Dimension,
+        "You need to provide the correct number of interpolation points");
+    static_assert(NumberOfVariablesToInterpolate <= NumberOfVariables,
+                  "You are trying to interpolate more variables than this "
+                  "container holds.");
+
+    auto weights = getWeights(targetPoints...);
+
+    std::array<double, NumberOfVariablesToInterpolate> interpolatedValues;
+
+    for (size_t nn = 0; nn < NumberOfVariablesToInterpolate; ++nn)
+      interpolatedValues[nn] = interpolate(weights, VariablesToInterpolate[nn]);
+
+    return interpolatedValues;
+  }
+
+  template <size_t NumberOfVariablesToInterpolate, size_t... I>
+  std::array<double, NumberOfVariablesToInterpolate> interpolate(
+      Index<NumberOfVariablesToInterpolate>& VariablesToInterpolate,
+      std::array<double, Dimension>& targetPoints,
+      std::index_sequence<I...>) const {
+    return interpolate(VariablesToInterpolate, targetPoints[I]...);
+  }
+
+  template <size_t NumberOfVariablesToInterpolate>
+  std::array<double, NumberOfVariablesToInterpolate> interpolate(
+      Index<NumberOfVariablesToInterpolate>& VariablesToInterpolate,
+      std::array<double, Dimension>& targetPoints) const {
+    return interpolate(VariablesToInterpolate, targetPoints,
+                       std::make_index_sequence<Dimension>{});
+  }
+
+  MultiLinearSpanInterpolation(
+      std::array<gsl::span<double const>, Dimension> x_,
+      gsl::span<double const> y_, Index<Dimension> numberOfPoints_)
+      : numberOfPoints(numberOfPoints_), x(x_), y(y_) {
+    for (size_t i = 0; i < Dimension; ++i) {
+      spacing[i] = x[i][1] - x[i][0];
+      inverseSpacing[i] = 1. / spacing[i];
+      allow_extrapolation_below_data[i] = false;
+      allow_extrapolation_above_data[i] = false;
+    };
+  };
+
+  void pup(PUP::er& p) {
+    //    p | x;
+    //    p | y;
+    p | numberOfPoints;
+    p | inverseSpacing;
+    p | allow_extrapolation_below_data;
+    p | allow_extrapolation_above_data;
+  }
+
+  const double& xmin(size_t whichDimension) const {
```

Comment:
maybe `lower_bound` and `upper_bound`? I don't feel strongly about this. General question about what you prefer.

`const size_t...` and below

Since the size of a pointer and `double` are the same we've always just returned `double` instead of `const double&`, which aids in eliminating a pointer chase.